modules/matchRoutes.js

```diff
@@ -41,9 +49,16 @@ function getIndexRoute(route, location, paramNames, paramValues, callback) {
       params: createParams(paramNames, paramValues)
     }
 
-    route.getIndexRoute(partialNextState, (error, indexRoute) => {
+    const indexRoutesReturn = route.getIndexRoute(partialNextState, (error, indexRoute) => {
       callback(error, !error && createRoutes(indexRoute)[0])
     })
+
+    if (isPromise(indexRoutesReturn))
```

Comment:
I've actually used that pattern to return a slightly different component resolved synchronously based on the query. I just invoked the cb synchronously though. Heh.